content/en/blog/_posts/2022-07-27-crd-validation-rules-graduate-to-beta.md

```diff
@@ -0,0 +1,146 @@
+| Constraint the max size limits on maps (objects with additionalProperties), arrays and string. | Use OpenAPIv3 value validations. Recommended for all maps, arrays and strings. This best practice is essential for rule cost estimation (explained below). | <pre>type:<br>maxItems: 100</pre> |
+| Require a date-time be more recent than a particular timestamp. | Use OpenAPIv3 string formats to declare that the field is a date-time. Use validation rules to compare it to a particular timestamp. | <pre>type: string<br>format: date-time<br>x-kubernetes-validations:<br>  - rule: "self >= timestamp('2000-01-01T00:00:00.000Z')"</pre> |
+| Require two sets to be disjoint. | Use x-kubernetes-list-type to validate that the arrays are sets. <br>Use validation rules to validate the sets are disjoint. | <pre>type: object<br>properties:<br>  set1:<br>    type: array<br>    x-kubernetes-list-type: set<br>  set2: ...<br>  x-kubernetes-validations:<br>    - rule: "!self.set1.all(e, !(e in self.set2))"</pre>
+
+## Using Transition Rules
+
+[Transition Rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#transition-rules) make it possible to compare the new state against the old state of a resource in validation rules.  Any validation rule that uses 'oldSelf' is a transition rule.  Transition rules are only evaluated when an old value and new value exist.
+
+Transition rule examples:
+
+| Transition Rule | Purpose |
+| - | - |
+| `self == oldSelf` | Validate that a required field is immutable once it is set. This does not prevent an optional field from transitioning from unset to set or set to unset.|
+| on parent of field: `has(self.field) == has(oldSelf.field)`<br>on field:  `self == oldSelf` | Validate that an optional field never changes after the resource is created.|
+| `self.all(x, x in oldSelf)` | Validate that a set is append-only.|
+| `self >= oldSelf` | Validate that a number is monotonically increasing.|
+
+
+## Using the Functions Libraries
+
+Validation rules have access to a couple different function libraries:
+
+* CEL standard functions, defined in the [list of standard definitions](https://github.com/google/cel-spec/blob/v0.7.0/doc/langdef.md#list-of-standard-definitions)
+* CEL standard [macros](https://github.com/google/cel-spec/blob/v0.7.0/doc/langdef.md#macros)
+* CEL [extended string function library](https://pkg.go.dev/github.com/google/cel-go@v0.11.2/ext#Strings)
+* Kubernetes [CEL extension library](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#pkg-functions) which includes supplemental functions for [lists](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#pkg-functions), [regex](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#Regex), and [URLs](https://pkg.go.dev/k8s.io/apiextensions-apiserver@v0.24.0/pkg/apiserver/schema/cel/library#Regex).
```

Comment:
Should these links actually be to v0.25.0?

Reply:
No, looks like I can omit the version part entirely. I'll fix.